cachito/web/content_manifest.py

```diff
@@ -0,0 +1,101 @@
+
+
+class ContentManifest:
+    """A content manifest associated with a Cacihto request."""
+
+    version = 1
+    json_schema_url = (
+        "https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/"
+        "f4abcfdaf8247a6b074f94fa84f3846f82d781c6/atomic_reactor/schemas/content_manifest.json"
+    )
+    unknown_layer_index = -1
+
+    def __init__(self, request=None):
+        """
+        Initialize ContentManifest.
+
+        :param Request request: the request to generate a ContentManifest for
+        """
+        self.request = request
+        self._packages = None
+        self._gopkg_data = {}
+        self._gomod_data = {}
+
+    def process_go_package(self, package, dependency):
+        """
+        Process go-package package.
+
+        :param Package package: the go-package package to process
+        :param Dependency dependency: the go-package package dependency to process
+        """
+        package = package.to_purl()
+        self._gopkg_data.setdefault(package, {"purl": package, "dependencies": [], "sources": []})
+        if dependency.type == "go-package":
+            icm_dependency = {"purl": dependency.to_purl()}
+            self._gopkg_data[package]["dependencies"].append(icm_dependency)
+
+    def process_gomod(self, package, dependency):
+        """
+        Process gomod package.
+
+        :param Package package: the gomod package to process
+        :param Dependency dependency: the gomod package dependency to process
+        """
+        self._gomod_data.setdefault(package.id, {"sources": [], "pkg": None})
+        if dependency.type == "gomod":
+            icm_source = {"purl": dependency.to_purl()}
+            self._gomod_data[package.id]["sources"].append(icm_source)
+        elif dependency.type == "go-package":
+            if self._gomod_data[package.id]["pkg"]:
+                raise RuntimeError(f'Multiple go-package dependencies for package "{package.id}"')
+            icm_package = dependency.to_purl()
+            self._gomod_data[package.id]["pkg"] = icm_package
+
+    def to_json(self):
+        """
+        Generate the JSON representation of the content manifest.
+
+        :return: the JSON form of the ContentManifest object
+        :rtype: dict
+        """
+        for req_dep in self.request.request_dependencies:
+            if req_dep.package.type == "gomod":
+                self.process_gomod(req_dep.package, req_dep.dependency)
+
+            elif req_dep.package.type == "go-package":
+                self.process_go_package(req_dep.package, req_dep.dependency)
+
+            else:
+                flask.current_app.logger.debug(
+                    "No ICM implementation for '%s' packages", req_dep.package.type
+                )
+
+        # gomod and go-package all refer to the gomod pkg manager. Let's merge their data
+        for k, v in self._gomod_data.items():
+            # v["pkg"] may be None for old DB entries from pre-content-manifest requests
```

Comment:
Good catch!

Reply:
We are now dealing with this differently: while we are still not generating ICMs for npm packages, gomod packages in requests without package level dependencies will have ICMs with only the "sources" field populated, i.e., only build time dependencies will be listed for those.